vendor/src/github.com/microsoft/hcsshim/createprocess.go

```diff
@@ -12,30 +14,71 @@ import (
 // processParameters is use to both the input of CreateProcessInComputeSystem
 // and to convert the parameters to JSON for passing onto the HCS
 type CreateProcessParams struct {
-	ApplicationName                   string
-	CommandLine                       string
-	WorkingDirectory                  string
-	StdInPipe, StdOutPipe, StdErrPipe string
-	Environment                       map[string]string
-	EmulateConsole                    bool
-	ConsoleSize                       [2]int
+	ApplicationName  string
+	CommandLine      string
+	WorkingDirectory string
+	Environment      map[string]string
+	EmulateConsole   bool
+	ConsoleSize      [2]int
+}
+
+// pipe struct used for the stdin/stdout/stderr pipes
+type pipe struct {
+	handle syscall.Handle
+}
+
+func makePipe(h syscall.Handle) *pipe {
+	p := &pipe{h}
+	runtime.SetFinalizer(p, (*pipe).closeHandle)
+	return p
+}
+
+func (p *pipe) closeHandle() {
+	if p.handle != 0 {
+		syscall.CloseHandle(p.handle)
+		p.handle = 0
+	}
+}
+
+func (p *pipe) Close() error {
+	p.closeHandle()
+	runtime.SetFinalizer(p, nil)
+	return nil
+}
+
+func (p *pipe) Read(b []byte) (int, error) {
+	// syscall.Read returns 0, nil on ERROR_BROKEN_PIPE, but for
+	// our purposes this should indicate EOF. This may be a go bug.
```

Comment:
Just by convention... The pipes are actually unidirectional, but you can't see how they're created since Windows creates them for us now. If you tried to call write on a reader pipe, you'd get some kind of error.

We always return the pipe as a io.Reader or io.Writer to restrict the caller to the correct direction, so I don't think this will be error prone.